Add the ability to load a checkpoint without restoring state

moinnadeem opened the pull request:
# Motivation
This pull request enables the ability to load a checkpoint without loading the associated state. It does so via the following YAML changes:
1. Introduces a `CheckpointLoaderHparams` object, with three fields: `checkpoint_filepath`, `load_weights_only`, and `strict`. 
2. If `load_weights_only = False`, then nothing changes and the previous codepaths are used. The `strict` value isn't considered if `load_weights_only = False`, since restoring a checkpoint with state should ensure that the model exactly matches up. YAML validation ensures that the `strict` value cannot be set without `load_weights_only = True`.
3. If `load_weights_only = True`, then it loads the checkpoint and avoids recovering the state via a new codepath. If `strict = False`, it also prints the keys that did not match up for user safety.

It creates the `CheckpointLoader` object when the `Trainer` is created via `create_from_hparams`, and passes the CheckpointLoader in as well.

# Discussion Points
1. Are we happy with the YAML API change?
2. Should we add any tests for these new codepaths?

ravi-mosaicml commented on composer/trainer/checkpoint_hparams.py:
```diff
@@ -0,0 +1,40 @@
+
+    See the documentation for the :class:`CheckpointLoader`.
+    """
+    checkpoint_filepath: Optional[str] = hp.optional(doc="Path to the serialized state_dict to recover state from.",
```
I think this should be required. The `CheckpointLoaderHparams` in the `TrainerHparms` can be none.

moinnadeem replied:
Agreed, changing this to be required.

moinnadeem replied:
This has been addressed!

ravi-mosaicml commented on composer/trainer/checkpoint_hparams.py:
```diff
@@ -0,0 +1,40 @@
+            return CheckpointLoader(checkpoint_filepath=self.checkpoint_filepath,
+                                    load_weights_only=self.load_weights_only,
+                                    strict=self.strict)
+        return None
```
This `return None` is probably causing a pyright bug.

moinnadeem replied:
Taken care of!

ravi-mosaicml commented on composer/trainer/checkpoint.py:
```diff
@@ -25,14 +26,29 @@ class CheckpointLoader:
-        state.load_state_dict(self.state_dict["state"])
+        if self.load_weights_only:
+
+            if self.state_dict['state']["_is_model_ddp_wrapped"] and not isinstance(state.model,
```
Can you refactor some of this logic out of state.load_state_dict so it's not duplicated? Perhaps create a method on the state class called `def load_model_state(self, state_dict: Dict[str, Any], strict: bool)`. Then the `load_state_dict` and state and this logic can call that.

moinnadeem replied:
Done!

ravi-mosaicml commented on composer/trainer/checkpoint.py:
```diff
@@ -25,14 +26,29 @@ class CheckpointLoader:
+    def __init__(self, checkpoint_filepath: str, load_weights_only: bool = False, strict: bool = False):
         self.state_dict = torch.load(checkpoint_filepath, map_location='cpu')
+        self.load_weights_only = load_weights_only
+        self.strict = strict
```
How would the strict parameter be used if `load_weights_only is False`?

moinnadeem replied:
I added validation, so `strict` cannot be used if `load_weights_only = False`

ravi-mosaicml commented on composer/trainer/checkpoint.py:
```diff
@@ -25,14 +26,29 @@ class CheckpointLoader:
         """
 
-        state.load_state_dict(self.state_dict["state"])
+        if self.load_weights_only:
```
To avoid duplicated logic, can you modify `State.load_state_dict` to take a `keys_to_load` parameter?

ravi-mosaicml reviewed (commented):
Looks pretty good overall, just some minor changes and comments.

hanlint commented on composer/trainer/trainer.py:
```diff
@@ -386,7 +384,7 @@ def create_from_hparams(cls, hparams: TrainerHparams) -> Trainer:
 
             # Checkpointing hparams
-            checkpoint_filepath=hparams.checkpoint_filepath,
+            checkpoint_loader=checkpoint_loader,
```
I find this API a bit confusing .. checkpoint_loader is an hparams file that has a `checkpoint_filepath`, but the root trainer hparams also has `checkpoint_folder`?

For consistency, I would suggest not having a spearate nested hparams object, and keeping all the settings flat.

moinnadeem replied:
Yes, this is because `CheckpointLoader` and `Checkpointer` are currently two different classes, where they should be one. So `checkpoint_folder` is for saving checkpoints, and `checkpoint_filepath` is for loading a checkpoint. 

To be clear, this is the previous API design, but I agree with your comments. 

@ravi-mosaicml thoughts if we should merge the hparams for saving and loading into a unified interface? 

This would broaden the scope of the PR (my goal was to get a blocker for BERT out of the way), but it probably is the correct thing to do.

moinnadeem replied:
> For consistency, I would suggest not having a spearate nested hparams object, and keeping all the settings flat

Do we think this adds too many hparams to the base YAML? I would almost prefer something like:
```
load_checkpoint: 
   filepath: str
   load_weights_only: bool
   strict: bool
save_checkpoint:
   interval_unit: str
   folder: str
   interval: int
```

If we keep them flat, then the `folder` and `filepath` can become confusing, whereas one requires a folder (for saving multiple checkpoints) and the other requires the link to a single file (for loading a single checkpoint). This also doesn't clutter the main Trainer interface.

Additionally, `save_checkpoint.interval` should use the `Time` abstraction when it is ready.

ravi-mosaicml replied:
I would prefer to have a `checkpoint` section in the haparms where you can then have that above yaml structure.

moinnadeem replied:
Yes, that is my thought. If both of you agree, I am happy to make the Trainer required changes. If we do that, can someone take care of addressing tests? I don't want to focus on this PR to the detriment of getting BERT ready in time.

ravi-mosaicml replied:
The save checkpoint won't require any additional tests. For load checkpoint with strict=False or load_weights_only=True, I think it would be trivial to add that in...I would be OK with a functional test (that just makes sure it runs, and doesn't necessarily assert that that warnings are raised or that only the model is loaded).

hanlint replied:
I'm OK with this approach! Yes maybe @ravi-mosaicml if you aren't overloaded can help get this merged?

moinnadeem replied:
@hanlint @ravi-mosaicml Awesome, I moved the checkpoint saving features into a `save_checkpoint` hyperparameter class. We now have two main level hparams: `load_checkpoint` and `save_checkpoint`, which have their respective hyperparameters.

moinnadeem commented:
Cool, addressed all feedback so far!

re: the PyRight issues, they're not problems in practice, should we add a manual ignore? In more detail:
```
 /home/runner/work/composer/composer/composer/trainer/checkpoint.py:47:60 - error: Argument of type "bool | None" cannot be assigned to parameter "strict" of type "bool" in function "load_model_state"
    Type "bool | None" cannot be assigned to type "bool"
      Type "None" cannot be assigned to type "bool" (reportGeneralTypeIssues)
```
The argument can't be None, because the method signature enforces a default. It seems as if PyRight isn't catching onto this?

```
/home/runner/work/composer/composer/composer/trainer/trainer.py
  /home/runner/work/composer/composer/composer/trainer/trainer.py:304:52 - error: "load_checkpoint" is not a known member of "None" (reportOptionalMemberAccess)
```
Before this line, we check if the `checkpoint_loader` is not None`, so `load_checkpoint` can never be run on an object of NoneType. Should we manually ignore this?

ravi-mosaicml commented:
For the pyright issues where it's complaining about optional variables, you need to add one of these options before the line it's complaining about:

1. `assert x is not None` (if it's an invariant violation)
2. `if x is None: raise ValueError(f"x is None, but it shouldn't be because ..."}` (if it's a user error)

ravi-mosaicml reviewed (commented):
Code path looks great! Please a some test to verify it. Thinking something like this:
1. Train, save checkpoint
2. Load checkpoint with a different optimizer and scheduler with `weights_only=True` 
3. Assert that the weights are the same as the first trainer, but that the optimizer is the new one

moinnadeem commented on composer/trainer/checkpoint_hparams.py:
```diff
@@ -0,0 +1,60 @@
+
+    def validate(self):
+        if not self.load_weights_only and self.strict:
+            raise ValueError(
```
Ah, good catch! I wrote this code on the plane, so it could have been a little better ;)

moinnadeem commented on composer/trainer/checkpoint_hparams.py:
```diff
@@ -0,0 +1,60 @@
+    filepath: str = hp.required(doc="Path to the serialized state_dict to recover state from.")
+    load_weights_only: Optional[bool] = hp.optional(doc="Whether to only load the weights from the model.",
+                                                    default=False)
+    strict: Optional[bool] = hp.optional(
```
@ravi-mosaicml How do you feel about `strict_weights`?

moinnadeem commented:
Cool, I've addressed all feedback and added the tests that Ravi requested. Hanlin also requested that, instead of passing the Checkpoint{Loader, Saver} object to the Trainer, I pass the hparams directly to make it more BYOT friendly. I agree there, so that has also been reflected.

I've addressed all PyRight issues on my changed files, but it seems as if PyRight is complaining about a few extra files that are outside of the scope of this PR. Namely, `composer/algorithms/augmix/augmix.py`, `composer/algorithms/randaugment/randaugment.py`, and `composer/datasets/brats.py`. What should we do about these?

ravi-mosaicml commented:
Can you merge in the latest from dev and see if that fixes those files?

moinnadeem commented:
@ravi-mosaicml Just did -- didn't help for some reason. Any clue why this is happening?

ravi-mosaicml commented:
Probably a pyright update...can you fix what's it complaining about in
those files if they're small changes?

On Mon, Jan 3, 2022, 7:22 AM Moin Nadeem ***@***.***> wrote:

> @ravi-mosaicml <https://github.com/ravi-mosaicml> Just did -- didn't help
> for some reason. Any clue why this is happening?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mosaicml/composer/pull/169#issuecomment-1004165347>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUYBL6H5SR4MF4PB5IPANTTUUG5MPANCNFSM5KHHTSGA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ravi-mosaicml reviewed (approved):
LGTM

ravi-mosaicml commented on composer/core/state.py:
```diff
@@ -213,7 +213,24 @@ def state_dict(self) -> types.StateDict:
         return state_dict
 
-    def load_state_dict(self, state: types.StateDict):
+    def load_model_state(self, state_dict: types.StateDict, strict: bool):
```
NVM, this doesn't make sense, since the function is only for the model.

jbloxham commented on tests/trainer/test_checkpoint.py:
```diff
@@ -109,6 +138,100 @@ def inject_stateful_callback_hparams(monkeypatch: MonkeyPatch):
+    # setup a new optimizer
+    second_trainer_hparams.optimizer = AdamWHparams()
+
+    # setup a new LR scheduler
+    scheduler_options = [ConstantLRHparams(), CosineAnnealingLRHparams(T_max=f"{second_trainer_hparams.max_epochs}ep")]
+    second_trainer_hparams.schedulers = [random.choice(scheduler_options)]
```
@moinnadeem just ran into this now - it's very strange to use randomness in a test in this way since it can potentially cause flaky tests. Was there a reason you wanted this randomness here?

moinnadeem replied:
Ah, I see the rationale. I just wanted to make sure that we test several schedulers, but didn't think it was worth the time to test all of them. In hindsight, we should either pick the more difficult one, or do both. I agree with you.

jbloxham replied:
I'm modifying this line in a PR for a different purpose anyways, so I'll fix it here.

moinnadeem replied:
Thanks Jamie!
